### ISSUE TITLE
Error/warning during Naive Bayes Training

### ISSUE THREAD

**Issue by bellafeng** [topepo/caret#793]
Hi, Max,
This is my first request. I apologize if I don't give enough information.
I am trying to replicate this from Machine Learning with R:
https://rpubs.com/jesuscastagnetto/caret-naive-bayes-spam-ham-sms
At the training stage, I got this. I saw somewhere online saying downgrade caret to a lower version would work, but I think it makes more sense to have the updated version work.
thanks, 
Bella
> sms_model1 <- train(sms_train, sms_raw_train$type, method="nb",   trControl=ctrl)
Something is wrong; all the Accuracy metric values are missing:
    Accuracy       Kappa    
 Min.   : NA   Min.   : NA  
 1st Qu.: NA   1st Qu.: NA  
 Median : NA   Median : NA  
 Mean   :NaN   Mean   :NaN  
 3rd Qu.: NA   3rd Qu.: NA  
 Max.   : NA   Max.   : NA  
 NA's   :2     NA's   :2    
Error: Stopping
In addition: There were 21 warnings (use warnings() to see them)
> sms_model1
Error: object 'sms_model1' not found

### My Session Info:
```{R}
>sessionInfo()

```
R version 3.4.2 (2017-09-28)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] klaR_0.6-12        MASS_7.3-47        dplyr_0.7.4        pander_0.6.1       caret_6.0-77       ggplot2_2.2.1     
 [7] lattice_0.20-35    e1071_1.6-8        wordcloud_2.5      RColorBrewer_1.1-2 tm_0.7-2           NLP_0.1-11        
[13] gmodels_2.16.2     class_7.3-14      

loaded via a namespace (and not attached):
 [1] ddalpha_1.3.1      tidyr_0.7.2        sfsmisc_1.1-1      splines_3.4.2      foreach_1.4.3      prodlim_1.6.1     
 [7] gtools_3.5.0       assertthat_0.2.0   stats4_3.4.2       DRR_0.0.2          yaml_2.1.14        robustbase_0.92-8 
[13] slam_0.1-40        ipred_0.9-6        backports_1.1.1    glue_1.2.0         digest_0.6.12      colorspace_1.3-2  
[19] recipes_0.1.1      htmltools_0.3.6    Matrix_1.2-11      plyr_1.8.4         psych_1.7.8        timeDate_3012.100 
[25] pkgconfig_2.0.1    CVST_0.2-1         broom_0.4.2        purrr_0.2.4        scales_0.5.0       gdata_2.18.0      
[31] gower_0.1.2        lava_1.5.1         combinat_0.0-8     tibble_1.3.4       withr_2.1.0        nnet_7.3-12       
[37] lazyeval_0.2.1     mnormt_1.5-5       survival_2.41-3    magrittr_1.5       evaluate_0.10.1    nlme_3.1-131      
[43] xml2_1.1.1         dimRed_0.1.0       foreign_0.8-69     tools_3.4.2        stringr_1.2.0      kernlab_0.9-25    
[49] munsell_0.4.3      bindrcpp_0.2       compiler_3.4.2     RcppRoll_0.2.2     rlang_0.1.4        grid_3.4.2        
[55] iterators_1.0.8    rmarkdown_1.6      gtable_0.2.0       ModelMetrics_1.1.0 codetools_0.2-15   reshape2_1.4.2    
[61] R6_2.2.2           lubridate_1.7.1    knitr_1.17         bindr_0.1          rprojroot_1.2      stringi_1.1.5     
[67] parallel_3.4.2     Rcpp_0.12.13       rpart_4.1-11       tidyselect_0.2.3   DEoptimR_1.0-8

You can delete the text in each section that explains how to do it correctly.
Be sure to test your 2 chunks of code in an empty R session before submitting your issue!

**Comment by bellafeng** [comment]
> warnings()
Warning messages:
1: predictions failed for Fold01: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

2: model fit failed for Fold01: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

3: predictions failed for Fold02: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

4: model fit failed for Fold02: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

5: predictions failed for Fold03: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

6: model fit failed for Fold03: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

7: predictions failed for Fold04: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

8: model fit failed for Fold04: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

9: predictions failed for Fold05: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

10: model fit failed for Fold05: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

11: predictions failed for Fold06: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

12: model fit failed for Fold06: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

13: predictions failed for Fold07: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

14: model fit failed for Fold07: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

15: predictions failed for Fold08: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

16: model fit failed for Fold08: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

17: predictions failed for Fold09: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

18: model fit failed for Fold09: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

19: predictions failed for Fold10: usekernel= TRUE, fL=0, adjust=1 Error in log(sapply(1:nattribs, tempfoo)) : 
  non-numeric argument to mathematical function

20: model fit failed for Fold10: usekernel=FALSE, fL=0, adjust=1 Error in `[<-`(`*tmp*`, , !num, value = 1) : 
  (subscript) logical subscript too long

21: In nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo,  ... :
  There were missing values in resampled performance measures.

**Comment by topepo** [comment]
It would be helpful to have a _small_ reproducible example. It is a bit much to comb through there blog post. Try it with a smaller subset of the data and send me the smaller data set once you can reproducible the issue. 

Also, that version of `caret` is from 2014-11-10 so I wouldn't be surprised if the code base has changed since then.

**Comment by bellafeng** [comment]
Hi, Max,
Thank you for the suggestion! I tried subsetting the original data to the first 100 records. the errors are gone. I increased the size to 1000, the model runs OK but with warnings like below. When I increase the size to 5000, the model doesn't run and I got errors (saying stopping) and warnings like before. 
So sorry, I can't reproduce it without sending the whole dataset and the code.
Warning messages:
1: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 1
2: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 3
3: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 4
4: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 6
5: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 8
6: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 9
7: In FUN(X[[i]], ...) :
  Numerical 0 probability for all classes with observation 11

**Comment by bellafeng** [comment]
Hi, Max,
Below is the code and attached is the whole dataset.
thanks!
Bella
# libraries needed by caret
library(klaR)
library(MASS)
# for the Naive Bayes modelling
library(caret)
# to process the text into a corpus
library(tm)
# to get nice looking tables
library(pander)
# to simplify selections
library(dplyr)


# a utility function for % freq tables
frqtab <- function(x, caption) {
  round(100*prop.table(table(x)), 1)
}
# utility function to summarize model comparison results
sumpred <- function(cm) {
  summ <- list(TN=cm$table[1,1],  # true negatives
               TP=cm$table[2,2],  # true positives
               FN=cm$table[1,2],  # false negatives
               FP=cm$table[2,1],  # false positives
               acc=cm$overall["Accuracy"],  # accuracy
               sens=cm$byClass["Sensitivity"],  # sensitivity
               spec=cm$byClass["Specificity"])  # specificity
  lapply(summ, FUN=round, 2)
}
# if (!file.exists("smsspamcollection.zip")) {
#   download.file(url="http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/smsspamcollection.zip",
#                 destfile="smsspamcollection.zip", method="curl")
# }
# sms_raw <- read.table(unz("smsspamcollection.zip","SMSSpamCollection"),
#                       header=FALSE, sep="\t", quote="", stringsAsFactors=FALSE)
sms_raw <- read.csv("C:/Users/gfeng/Documents/_datascience/Machine-Learning-with-R-datasets-master/sms_spam.csv", stringsAsFactors=FALSE)
sms_raw <- sms_raw[1:5000, ]
colnames(sms_raw) <- c("type", "text")
sms_raw$type <- factor(sms_raw$type)
# randomize it a bit
set.seed(12358)
sms_raw <- sms_raw[sample(nrow(sms_raw)),]
str(sms_raw)

sms_corpus <- Corpus(VectorSource(sms_raw$text))
sms_corpus_clean <- sms_corpus %>%
  tm_map(content_transformer(tolower)) %>% 
  tm_map(removeNumbers) %>%
  tm_map(removeWords, stopwords(kind="en")) %>%
  tm_map(removePunctuation) %>%
  tm_map(stripWhitespace)
sms_dtm <- DocumentTermMatrix(sms_corpus_clean)

train_index <- createDataPartition(sms_raw$type, p=0.75, list=FALSE)
sms_raw_train <- sms_raw[train_index,]
sms_raw_test <- sms_raw[-train_index,]
sms_corpus_clean_train <- sms_corpus_clean[train_index]
sms_corpus_clean_test <- sms_corpus_clean[-train_index]
sms_dtm_train <- sms_dtm[train_index,]
sms_dtm_test <- sms_dtm[-train_index,]

ft_orig <- frqtab(sms_raw$type)
ft_train <- frqtab(sms_raw_train$type)
ft_test <- frqtab(sms_raw_test$type)
ft_df <- as.data.frame(cbind(ft_orig, ft_train, ft_test))
colnames(ft_df) <- c("Original", "Training set", "Test set")
pander(ft_df, style="rmarkdown",
       caption=paste0("Comparison of SMS type frequencies among datasets"))
sms_dict <- findFreqTerms(sms_dtm_train, lowfreq=5)
sms_train <- DocumentTermMatrix(sms_corpus_clean_train, list(dictionary=sms_dict))
sms_test <- DocumentTermMatrix(sms_corpus_clean_test, list(dictionary=sms_dict))

# modified sligtly fron the code in the book
convert_counts <- function(x) {
  x <- ifelse(x > 0, 1, 0)
  x <- factor(x, levels = c(0, 1), labels = c("Absent", "Present"))
}
sms_train <- sms_train %>% apply(MARGIN=2, FUN=convert_counts)
sms_test <- sms_test %>% apply(MARGIN=2, FUN=convert_counts)

ctrl <- trainControl(method="cv", 10)
set.seed(12358)
sms_model1 <- train(sms_train, sms_raw_train$type, method="nb",
                    trControl=ctrl)
sms_test_pred <- predict(sms_model1, sms_test)
library(gmodels)
CrossTable(sms_test_pred, sms_raw_test$type,
           prop.chisq = FALSE, prop.t = FALSE,
           dnn = c('predicted', 'actual'))

[sms_spam.zip](https://github.com/topepo/caret/files/1504755/sms_spam.zip)

**Comment by topepo** [comment]
I'll look at this. Those warnings are not a big deal; one of you new data points falls outside of the range of one of the distributions (probably the conditional). There's not much to do about that.

**Comment by bellafeng** [comment]
Thank you, Max. The warnings are OK. But when I increase the size of the dataset to over 5000, errors started and the model wouldn't be created, like the following:
sms_model1 <- train(sms_train, sms_raw_train$type, method="nb", trControl=ctrl)
Something is wrong; all the Accuracy metric values are missing:
Accuracy Kappa
Min. : NA Min. : NA
1st Qu.: NA 1st Qu.: NA
Median : NA Median : NA
Mean :NaN Mean :NaN
3rd Qu.: NA 3rd Qu.: NA
Max. : NA Max. : NA
NA's :2 NA's :2
Error: Stopping
In addition: There were 21 warnings (use warnings() to see them)
sms_model1
Error: object 'sms_model1' not found

**Comment by bellafeng** [comment]
oh, just realized what you meant. so I will remove that data points and try later. thank you, Max!

**Comment by bellafeng** [comment]
Hi, Max,
For some weird reason, when I remove this record from the csv file, the error is gone.
ham	Donâ€˜t give a flying monkeys wot they think and I certainly donâ€˜t mind. Any friend of mine and all that!
I suspected it might be because of the "â€˜" unicode (but it appears in other places as well). 
thank you, 
Bella

**Comment by Mihiretukebede** [comment]
I am running in the same error. I used the same article to replicate in my dataset
I have a large biomedical literature reference data that I read using the *Bibliometrix* package. I followed the same way it was described in the article referenced by @bellafeng to classify abstracts. I would like to use the mighty *caret* package since it gives me opportunities to control my tuning parameters. 
 
`set.seed(12358)
ctrl <- trainControl(method="cv", 10)

abs_model11 <- train(abs_train, ab_train$Include, method="nb",
                trControl=ctrl)
abs_model1`

**Result**
Warning message:
"predictions failed for Fold01: usekernel= TRUE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold01: usekernel=FALSE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold02: usekernel= TRUE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold02: usekernel=FALSE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold03: usekernel= TRUE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold03: usekernel=FALSE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
"predictions failed for Fold04: usekernel= TRUE, fL=0, adjust=1 Error in predict.NaiveBayes(modelFit, newdata) : 
  Not all variable names used in object found in newdata
"Warning message:
.
.
.

When I used a different model 

`set.seed(12358)
abs_nb_model <- naiveBayes(abs_train, ab_train$Include, laplace = 1,  
  type = c("class", "raw"))
abs_nb_model` # This worked 

#Prediction 
`nb_predict1 <- predict(abs_nb_model, abs_test)
cm1 <- confusionMatrix(as.factor(nb_predict1), as.factor(ab_test$Include), positive= "included")
cm1`

**Result**
Confusion Matrix and Statistics

          Reference
Prediction excluded included
  excluded     2845       13
  included     1825      245
                                          
               Accuracy : 0.627           
                 95% CI : (0.6134, 0.6406)
    No Information Rate : 0.9476          
    P-Value [Acc > NIR] : 1               
                                          
                  Kappa : 0.1294          
                                          
 Mcnemar's Test P-Value : <2e-16          
                                          
            Sensitivity : 0.94961         
            Specificity : 0.60921         
         Pos Pred Value : 0.11836         
         Neg Pred Value : 0.99545         
             Prevalence : 0.05235         
         Detection Rate : 0.04972         
   Detection Prevalence : 0.42005         
      Balanced Accuracy : 0.77941         
                                          
       'Positive' Class : included

**Comment by Mihiretukebede** [comment]
There is a match between the testing and training set features. The model works well without cross-validation and parameter tuning. However, there seems a mismatch in features for the cross-validation samples. No idea how to control that in the trainControl function. I thought trainControl do the cross-validation samples using observations without affecting the features.  
train_control <- trainControl(method = "cv", number = 10). 
I set sampling = "down" to check if the error was due to the class imbalance(95%, 5%). Same error.
